addons/mail/models/mail_thread.py

```diff
@@ -3648,25 +3648,35 @@ def _notify_thread_by_web_push(self, message, recipients_data, msg_vals=False, *
         """
         msg_vals = dict(msg_vals or {})
         partner_ids = self._extract_partner_ids_for_notifications(message, msg_vals, recipients_data)
-        if not partner_ids:
-            return
-
-        partner_devices_sudo = self.env['mail.push.device'].sudo()
-        devices = partner_devices_sudo.search([
-            ('partner_id', 'in', partner_ids)
-        ])
+        devices, private_key, public_key = self._get_web_push_parameters(partner_ids)
         if not devices:
             return
+        payload = self._truncate_payload(self._notify_by_web_push_prepare_payload(message, msg_vals=msg_vals))
+        self._push_web_notification(payload, devices, private_key, public_key)
 
+    def _get_web_push_parameters(self, partner_ids):
```

Comment:
Does it need to be on thread? I don't think it uses self at all.
Could be moved to tools?

Reply:
Keep it here. Otherwise you just end up jumping between files for 10 lines.